smarts/core/default_map_builder.py

```diff
@@ -23,11 +23,61 @@
 from typing import NamedTuple, Tuple
 from pathlib import Path
 from smarts.core.road_map import RoadMap
-from smarts.core.utils.file import file_md5_hash
+from smarts.core.utils.file import file_md5_hash, path2hash
+
 
 _existing_map = None
 
 
+def _cache_result(map_spec, road_map, road_map_hash: str):
+    global _existing_map
+    from smarts.sstudio.types import MapSpec
+
+    class _RoadMapInfo(NamedTuple):
+        map_spec: MapSpec
+        obj: RoadMap
+        map_hash: str
+
+    _existing_map = _RoadMapInfo(map_spec, road_map, road_map_hash)
+
+
+def _clear_cache():
+    global _existing_map
+    if _existing_map:
+        import gc
+
+        # Try to only keep one map around at a time...
+        del _existing_map
+        _existing_map = None
+        gc.collect()
+
+
+_UNKNOWN_MAP = 0
+_SUMO_MAP = 1
+_OPENDRIVE_MAP = 2
+
+
+def _find_mapfile_in_dir(map_dir: str) -> Tuple[int, str]:
+    map_filename_type = {
+        "map.net.xml": _SUMO_MAP,
+        "shifted_map-AUTOGEN.net.xml": _SUMO_MAP,
+        "map.xodr": _OPENDRIVE_MAP,
+    }
+    map_type = _UNKNOWN_MAP
+    map_path = map_dir
+    for f in os.listdir(map_dir):
+        cand_map_type = map_filename_type.get(f)
+        if cand_map_type is not None:
+            return cand_map_type, os.path.join(map_dir, f)
+        if f.endswith(".net.xml"):
+            map_type = _SUMO_MAP
+            map_path = os.path.join(map_dir, f)
+        elif f.endswith(".xodr"):
+            map_type = _OPENDRIVE_MAP
+            map_path = os.path.join(map_dir, f)
```

Comment:
If we have two files ending with `.net.xml`, wouldn't it add to the path creating a invalid path. I can see it happening for a SUMO scenario having the shifted map offset and the original network file.

Reply:
I'm not sure I understand this one correctly, but `map_path` is set "fresh" each time (not concatenated) from `map_dir`, so it doesn't "add to the path".

If there are both `map.net.xml` and `shifted_map-AUTOGEN.net.xml` in the same folder, the first one that is encountered (which should be `map.net.xml` if they are listed alphabetically, but it doesn't matter either way) will be returned on line 71 and the `map_type` will be SUMO.  If there is _also_ some other file in the folder, say `abc.net.xml`, it won't be returned, it will just cause the local `map_path` to be set and then discarded upon line 71.